Shopify/cli
# [Prototype] Versioned App Configuration

Based on [Support app config specifications for configuring app access](https://github.com/Shopify/cli/pull/2802/commits)

Co-authored-by: Alvaro Gutierrez <alvaro.gutierrezblanco@shopify.com>
Co-authored-by: Trish Ta <trish.ta@shopify.com>
Co-authored-by: Bill Fienberg <bill.fienberg@shopify.com>



### WHY are these changes introduced?

Fixes #0000 



### WHAT is this pull request doing?

- Creates the new `ConfigExtensionSpec`. It's a simplified version of the current `ExtensionSpec`. New management experience `app_config` remote specs will be mapped to this new `ConfigExtensionSpec` type.
- Modify `app loader` logic to use an app schema generated dynamically from the union of the subschema of each `app_config` extension spec
- Modify `app loader` logic to load each section of the app toml as a new `configExtensions` list inside the `app` object
- Support current toml format. Config specs support declarative transform configurations (ideally this should be done in the server side in the near future)
- Deploy command will include the current `allExtensions` list and the new `configExtensions` list
- `link` command build the app config toml from the `app_config` extension registrations. Added support for apps created with previous cli versions that don't have any `app_config` extension registrations.
- Refactored `release` and `deploy` command prompts. Now the content is build outside and they use the same prompt:
    - `deploy` - Build content based on local state and remote extension reigistrations active_version
    - `release` - Uses the appDiff API
- Included [new design](https://www.figma.com/file/cTu9Ag52V4IdqZe1cij6kK/Versioned-app-config?type=design&node-id=9657-5889&mode=design&t=lXjWi9wshavJk5Ch-0) for the `release` and `deploy` prompts 

### How to test your changes?
- Create a new spin instance `spin up partners -c shopify.branch=prototype-versioned-app-config -c partners.branch=add-app-config-management` 
- Create a new local app
- Run the link command and create a new remote app
- Create some extensions
- Modify different values in the `toml` file.
- Run `deploy` command.
- Run link command again and create different remote apps. 
- Add/delete extensions. Modify the toml and run link against different remote apps to check that the deployed content for each app is restored properly

### Post-release steps



### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
